modules/database/export.bal

```diff
@@ -0,0 +1,485 @@
+    TimelineSubstepExportBase[] substepsBase = check getTimelineSubstepsBaseExport(stepId);
+    foreach var substep in substepsBase {
+        SubstepDataExport[] substepDataList = check getTimelineSubstepDataList(stepId, substep.substepNr);
+        TimelineSubstepExport substepExport = {
+            substepId: substep.substepId,
+            href: substep.href,
+            hrefUi: substep.hrefUi,
+            cleared: substep.cleared,
+            parameters: substep.parameters,
+            parametersContentType: substep.parametersContentType,
+            substepDataList: substepDataList
+        };
+        substepsExport.push(substepExport);
+    }
+    return substepsExport;
+}
+
+public isolated transactional function getExperimentDataExport(int experimentId, int dataId) returns ExperimentDataExport|error {
+    stream<ExperimentDataExport, sql:Error?> experimentDataResult = experimentDB->query(
+        `SELECT dataId, name, version, location, type, contentType FROM ExperimentData WHERE dataId=${dataId} AND experimentId=${experimentId};`
+    );
+    var experimentData = experimentDataResult.next();
+    check experimentDataResult.close();
+    if experimentData is error {
+        return experimentData;
+    }
+    if experimentData is record {ExperimentDataExport value;} {
+        return experimentData.value;
+    } else {
+        // should never happen based on the sql query
+        return error(string `[experimentId ${experimentId}, dataId ${dataId}] Could not retrieve experiment data for export from database.`);
+    }
+}
+
+public isolated transactional function getExperimentDBExport(int experimentId) returns ExperimentCompleteExport|error {
+    TimelineStepExport[] timelineSteps = [];
+    ExperimentDataExport[] experimentDataList = [];
+    ExperimentFull experiment = check getExperiment(experimentId);
+    int[] dataIdList = [];
+
+    // iterate over timeline steps
+    TimelineStepFull[] timelineStepListDb = check getTimelineStepListExport(experimentId);
+    foreach TimelineStepFull timelineStepDb in timelineStepListDb {
+        TimelineStepExport timelineStepExport = check castToTimelineStepExport(timelineStepDb);
+        // retrieve associated step data
+        timelineStepExport.stepDataList = check getTimelineStepDataList(timelineStepDb.stepId);
+        foreach var stepData in timelineStepExport.stepDataList {
+            dataIdList.push(stepData.dataId);
+        }
+        // retrieve associated substeps with their substep data
+        timelineStepExport.timelineSubsteps = check getTimelineSubstepsExport(timelineStepDb.stepId);
+        timelineSteps.push(timelineStepExport);
+        foreach var substep in timelineStepExport.timelineSubsteps {
+            foreach var substepData in substep.substepDataList {
+                dataIdList.push(substepData.dataId);
+            }
+        }
+    }
+
+    // ignore duplicates in dataIdList and create experiment data list
```

Comment:
This screams set (at least the duplicates filtering part...). If you already retreive all data in one query, then you could export data based on if the id is in the set. It is so very unfortunate that ballerina only has a (limited, as in string keys only) map datatype... So I guess this can stay as is.